[iOS] 새로운 카드 추가 기능 구현

seizze opened the pull request:
### 구현 내용

* 스토리보드로 새 카드를 추가하는 Form 화면을 만들고, 스크롤뷰를 적용하여 키보드에 반응하도록 구현하였습니다.
* 서버에 새 카드 추가 요청을 하고, 새 카드의 ID를 응답받도록 구현하였습니다. (Mock Data를 이용중입니다.)
* 반환받은 새 카드 ID와 Form 화면을 통해 제출된 정보로 새 카드 인스턴스를 만들어 추가하고, 뷰에 나타내었습니다.

### 고민한 점

* 테이블뷰에서 바뀐 행만 업데이트하기 위해, 뷰모델에서 데이터가 업데이트 됐을 때 changeHandler에 카드 리스트만 넘겨주는 것이 아니라 리스트와 함께 변경된 행을 ListChangeDetail 타입으로 넘겨주도록 구현하였습니다. 테이블 뷰를 업데이트하는 곳에선 변경된 행이 있을 경우 해당 행만 업데이트하고, 아닐 경우에만 전체 reload를 하도록 구현하였습니다.
* 서버에서 응답받는 Response 모델들의 기본 기능을 나타내는 Response 프로토콜을 추가하고 extension을 통해 기본 기능을 구현하였습니다. 이 extension에 서버에서 응답받을 때 상태 코드의 유효성을 검사할 수 있는 `validate(statusCode:block:)`메서드를 추가하였습니다.

* 새 카드를 추가하는 Form 화면에서 키보드에 반응하도록 하기 위해 스크롤뷰를 추가하고, observer를 등록 후에 키보드가 올라가거나 내려갔을 때 스크롤뷰의 content inset을 조정하도록 구현하였습니다. 또한 현재 활성화된 control을 찾고 `scrollRectToVisible` 메서드를 이용하여 해당 control이 화면에 잘 보이도록 스크롤하였습니다.

### 궁금한 점

새 카드를 추가할 때, 정보의 흐름을 보면
* Form View Controller에서 제출 버튼을 누르면, Board View Controller에 delegation으로 전달
* Board VC에서 서버에 새 카드 ID를 요청하고, 응답 받으면 Notification 발송
* CardList VC에서 Notification을 받아서 뷰모델 업데이트

이런 식으로 전체적으로 돌아돌아 가는 느낌인데, 이렇게 해도 되는 것인지 잘 모르겠습니다.

저는 네트워킹 관련 코드가 분산되어 있지 않고 Board View Controller에 모여있는 편이 관리하기 쉽다고 생각해서 이렇게 했는데, Form VC에서 바로 네트워크 요청 후 Notification을 발송하는 편이 좋은가 하는 생각도 들어서 질문드립니다.

Related issues: #23, #24, #26 and #34

seizze committed:
[iOS] feat : Mock JSON 데이터 추가

associate with #24

seizze committed:
[iOS] feat : 카드 리스트 정보 업데이트

Closes #24

seizze committed:
[iOS] feat : 새 VC 인스턴스 추가

associate with #23

seizze committed:
[iOS] feat : FormViewController 파일 추가

associate with #23

seizze committed:
[iOS] feat : Form View Controller를 모달로 present

associate with #23

seizze committed:
[iOS] feat : 스토리보드 Form 화면 구성

associate with #23

seizze committed:
[iOS] feat : Form 화면을 스크롤뷰에 임베드

associate with #23

seizze committed:
[iOS] feat : 새 카드 이름 레이블을 텍스트필드로 변경

associate with #23

seizze committed:
[iOS] feat : 키보드에 따라서 스크롤뷰 조정

associate with #23

seizze committed:
[iOS] feat : 키보드에 따라서 텍스트뷰가 잘 보이도록 스크롤

associate with #23

seizze committed:
[iOS] feat : 취소 버튼 터치 시 dismiss

Closes #23

seizze committed:
[iOS] feat : 딜리게이트로 BoardVC에서 정보를 받도록 구현

associate with #26

seizze committed:
[iOS] feat : 새 카드 ID response 모델 추가

associate with #26

seizze committed:
[iOS] feat : API Router에서 request 반환하도록 변경

- 새 카드 추가 시 post로 요청해야 하므로 API Router에서 http method를 포함하여 request를 만들도록 변경

Associate with #26

seizze committed:
[iOS] feat : 서버에 새 카드 ID 요청후 응답받는 기능 구현

Associate with #26

seizze committed:
[iOS] refactor : Response 프로토콜 기능을 새 파일로 이동

Associate with #26

seizze committed:
[iOS] feat : 새 카드 ID 디코딩 완료

Associate with #26

seizze committed:
[iOS] feat : 새 카드 인스턴스를 노티로 post 구현

Associate with #26

seizze committed:
[iOS] feat : 노티를 받아 리스트 모델에 추가하도록 구현

Associate with #26

seizze committed:
[iOS] feat : 카드 인스턴스에 내용을 담음

Associate with #26

seizze committed:
[iOS] feat : 테이블 뷰의 추가된 부분만 업데이트하도록 개선

- 테이블 뷰를 전체 reload하지 않고, 뷰모델에서 추가된 row를 알려줘서 해당 row만 업데이트하도록 개선

Associate with #26

seizze committed:
[iOS] refactor : 필요 없는 함수 삭제

Associate with #26

seizze committed:
[iOS] docs : iOS 진행 상황 업데이트

Closes #26

godrm commented on mobile/TodoList/TodoList/Controllers/BoardViewController.swift:
```diff
@@ -51,13 +52,43 @@ extension BoardViewController {
+            case let .success(response):
+                guard let response = response as? CardIDResponse else { return }
+                card.id = response.cardID
+                center.post(name: .newCardDidUpdate, object: self, userInfo: card.package(listID: id))
```
여기 있는 두 requestBoard() requestNewCard() 에서 보내는 post()가 어색합니다.
네트워크에서 응답이 오면 뷰 컨트롤러가 보내고 있는데 
그 값이 저장되어야 할 모델이 업데이트되면 모델이 보내는 게 어떨까요?

이렇게 생각해보세요.
네트워크를 통해서 새로운 Card가 생긴 경우와 사용자가 새 카드를 만든 경우가 
동일한 흐름으로 동작하도록 해야 합니다. 지금처럼 하면 양쪽 모두에서 post를 할까요?

seizze replied:
네트워크 응답이 오면 하위 뷰 컨트롤러에게 데이터를 전달하고, 하위 뷰 컨트롤러에서는 데이터를 넘겨받아 뷰모델을 새로 생성하도록 변경하였습니다!
아직까지는 모델이 post할 일이 생기지 않아서 보내지 않았는데, post할 일이 생기면 모델이 보내도록 하겠습니다!!

godrm commented on mobile/TodoList/TodoList/Controllers/BoardViewController.swift:
```diff
@@ -51,13 +52,43 @@ extension BoardViewController {
+extension BoardViewController: CardListViewControllerDelegate {
+    func addNewCardDidTouch(listID id: Int?) {
+        guard let vc = UILoader.load(viewControllerType: FormViewController.self, from: storyboard) else { return }
+        vc.listID = id
```
이 값이 꼭 있어야 할까요? 어차피 아래에서 델리게이트 메소드로 넘겨 받지 않나요?

아.. 그리고보니 Delegate를 만들 때는 1:n을 고려해야 합니다. 
여러 FormViewController를 생성하는 곳이 여러 곳일 수도 있고, 한 곳일 수도 있습니다. 
그래서 델리게이트 프로토콜에서 메소드 첫 번째 인자값을 이벤트를 발생하는 객체가 들어있습니다.
AppDelegate, UITableViewDelegate, TextFieldDelegate 프로토콜을 보세요.

seizze replied:
넵 말씀하신 대로 어떤 객체에서 이벤트를 발생시켰는지를 알기 위해 델리게이트 메서드의 첫 번째 인자로 이벤트를 발생시킨 객체를 전달하도록 변경하였습니다.

godrm commented on mobile/TodoList/TodoList/Controllers/BoardViewController.swift:
```diff
@@ -51,13 +52,43 @@ extension BoardViewController {
+extension BoardViewController: FormViewControllerDelegate {
+    func newCardDidSubmit(listID id: Int?, card: Card) {
+        guard let id = id else { return }
+        requestNewCard(listID: id, card: card)
```
제 의견은 FormViewController는 id가 없어도 여기서 받으면 id를 이미 알고 있지 않나요?

seizze replied:
음 어떤 카드 리스트의 + 버튼을 눌렀는지를 BoardViewController에 전달하고, BoardViewController에서 FormViewController를 present한 뒤에 사용자가 FormViewController의 필드를 모두 쓰고 제출 버튼을 눌렀을 때 어떤 카드 리스트에 카드를 추가할 지를 알기 위해 FormViewController에 listID 프로퍼티를 갖고 있도록 했었습니다.

고민해 봤는데 id 프로퍼티 없이 여기서 id를 알 수 있는 방법을 잘 모르겠습니다 🤯

godrm commented on mobile/TodoList/TodoList/Controllers/CardListViewController.swift:
```diff
@@ -33,15 +41,27 @@ class CardListViewController: UIViewController {
+        
+        observer = Card.addCardObserver(forName: .newCardDidUpdate, listID: listID) { [weak self] in
+            self?.viewModel?.insert(card: $0)
+        }
```
여기 옵저버도 다시 생각해보세요. 오히려 모델이 뷰 모델이 업데이트되면 VC에 이벤트를 보내야 하는 거 아닐까요?

seizze replied:
넵 위에서 말씀드린 것처럼 네트워크 응답이 오면 하위 뷰 컨트롤러에게 데이터를 전달하고, 하위 뷰 컨트롤러에서는 데이터를 넘겨받아 뷰모델을 새로 생성하거나 업데이트 하도록 만들었더니 옵저버가 필요가 없어져서 이 부분은 지웠습니다!

godrm commented on mobile/TodoList/TodoList/Controllers/CardListViewController.swift:
```diff
@@ -65,3 +89,13 @@ private extension List {
+                                using block: @escaping (Card) -> Void) -> NSObjectProtocol {
+        return NotificationCenter.default.addObserver(forName: name, object: nil, queue: .main) {
+            if let card = $0.userInfo?[id] as? Card { block(card) }
+        }
```
옵저버를 추상화해서 바인드처럼 쓰는 것 같습니다. List 와 Card 모두 같은 구현을 써도 되겠는데요? 
제네릭을 배웠으니 다시 한 번 고민해보세요. 
Card나 List가 바뀌었을 때 받아야 하는 대상 옵저버가 1:n 관계가 맞나요? 
1:1 이라면 꼭 노티가 아니어도 다른 방식도 고려해보세요.

seizze replied:
이 부분도 옵저버가 필요 없어져서 지웠습니다. ㅎㅎ 다음에 필요해졌을 때 제네릭을 이용해서 개선해 보겠습니다!!

godrm commented on mobile/TodoList/TodoList/Models/Observers.swift:
```diff
@@ -25,3 +25,25 @@ class Observers {
+            block()
+        }
+        observers.append(observer)
+    }
```
사실 옵저버를 추가하면 객체 내부에 보관(?)이 되긴 합니다. 별도로 꼭 관리해야 하는 이유가 있을까요? 
그냥 복잡해보여서 추상화를 한건가요?

seizze replied:
음 addObserver에서 반환받은 값을 이용해서 VC deinit 시에 removeObserver를 호출하여 옵저버를 지워줘야 한다고 읽어서 나중에 옵저버를 제거해 주기 위해 저장해 두었었습니다. 이렇게 할 필요가 없나요...?

참고한 문서: [addObserver(forName:object:queue:using:)](https://developer.apple.com/documentation/foundation/notificationcenter/1411723-addobserver)

godrm commented on mobile/TodoList/TodoList/Models/Response.swift:
```diff
@@ -0,0 +1,38 @@
+
+enum StatusCode: String, Decodable {
+    case ok = "200"
+    case error = "400"
```
너무 성공과 실패로만 나눈것 같은데요 :)
URLResponse 나 HTTPURLResponse 에 이미 선언되어 있는건데 안 쓰고 분리한 이유가 궁금합니다.

seizze replied:
앗.. HTTPURLResponse에 있는 줄 몰랐습니다.

URLProtocolMock에 Response를 전달해주는 코드를 추가하였고, HTTPURLResponse extension을 이용해 statusCode가 200대인지를 확인하는 메서드를 추가하였습니다!

godrm commented on mobile/TodoList/TodoList/Networking/APIRouter.swift:
```diff
@@ -9,23 +9,63 @@
+            return nil
+        }
+        return data
+    }
```
APIRouter 객체는 다음 다른 프로젝트하면 재사용이 가능할까요? 
네트워크나 모델 처럼 하위 객체들은 그 객체만 분리해서 재사용이 가능한가 고려해보면 좋습니다.

seizze replied:
다른 프로젝트에서 재사용이 가능한 부분을 프로토콜과 익스텐션으로 분리하였습니다! 재사용이 불가능하고 투두 앱에 의존적인 부분은 프로토콜을 채택하여 구현하는 방식으로 해보았습니다.

godrm commented on mobile/TodoList/TodoList/NetworkingTests/JSONMocks/BoardResponseJSONMock.swift:
```diff
@@ -0,0 +1,108 @@
+        }]
+    }
+}
+""".data(using: .utf8)!
```
이런 데이터 파일은 그냥 별도 파일로 분리해도 좋습니다. 참고하세요.

seizze replied:
넵 .json 파일로 분리하고 `Bundle.main.url()` 메서드를 통해 불러와서 사용하도록 변경하였습니다!

seizze committed:
[iOS] refactor : 델리게이트 메서드 개선

- 델리게이트 메서드의 첫 번째 인자값으로 이벤트를 발생시긴 객체를 전달하도록 개선
Associate with #34

seizze committed:
[iOS] refactor : 네트워크 응답 후 뷰 업데이트 과정 개선

Associate with #34

seizze committed:
[iOS] refactor : 카드 추가 시 뷰 업데이트 과정 개선

Associate with #34

seizze committed:
[iOS] refactor : 프로토콜로 VC간 통신 연결

Associate with #34

seizze committed:
[iOS] refactor : URLProtocolMock에서 response 전달

Associate with #34

seizze committed:
[iOS] refactor : HTTPURLResponse을 사용하여 validate

Associate with #34

seizze committed:
[iOS] refactor : API Router의 일반적인 부분을 분리

Associate with #34

seizze committed:
[iOS] refactor : Mock json을 파일로 관리

Associate with #34

seizze committed:
[iOS] test : 변경된 뷰모델에 맞게 테스트 수정

Closes #34

seizze commented:
피드백 반영 완료하였습니다.
변경된 뷰모델에 맞게 테스트 코드도 변경하였습니다!
